<fix>[volume]: delete volume snapshot by block-commit

zstack-robot-1 opened the pull request:
APIImpact

Resolves: ZSV-5799

Change-Id: I6b72706a71796c646567686b746d756478656a78

sync from gitlab !8275

taogan21 committed:
<fix>[volume]: delete volume snapshot by block-commit

APIImpact

Resolves: ZSV-5799

Change-Id: I6b72706a71796c646567686b746d756478656a78

taogan21 committed:
<fix>[kvm]: update grayUpgrade.json

Resolves: ZSTAC-1000

Change-Id: I6f6b646c797674616363707661626370736b7172

zstack-robot-2 commented on header/src/main/java/org/zstack/header/storage/primary/PullVolumeSnapshotOnPrimaryStorageMsg.java:
```diff
@@ -0,0 +1,48 @@
+        this.dstSnapshot = dstSnapshot;
+    }
+    @Override
+    public String getPrimaryStorageUuid() {
+        return volume.getPrimaryStorageUuid();
+    }
```
**Comment from tao.gan:**

添加检测 volume 是否为 null

taogan21 committed:
<fix>[storage]: modify based on AI comments

1 Commit/Pull/Take/Merge all adopt the synchronization level specified by KVMGlobalConfig.HOST_SNAPSHOT_SYNC_LEVEL

2 Snapshot deletion logic: When the scope is equal to 'single',
the logic for deleting a single snapshot is executed;
when the scope is not equal to 'single',
the logic for deleting the snapshot chain is executed.

Resolves: ZSTAC-1000

Change-Id: I666671736e6e746f6e69746567746a6b656f746e

zstack-robot-1 commented on header/src/main/java/org/zstack/header/storage/snapshot/VolumeSnapshotDeletionStructs.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.List;
+
+public class VolumeSnapshotDeletionStructs {
+    private List<VolumeSnapshotInventory> snapshotInventories;
+    private String direction;
+    private String scope;
+
+    public VolumeSnapshotDeletionStructs(List<VolumeSnapshotInventory> snapshotInventories, String direction, String scope) {
+        this.snapshotInventories = snapshotInventories;
+        this.direction = direction;
+        this.scope = scope;
+    }
```
**Comment from tao.gan:**

```diff
private String direction;
private String scope;
```
作为参数是符合预期的，。不用修改 参数类型。

zstack-robot-2 commented on plugin/kvm/src/main/java/org/zstack/kvm/KVMHost.java:
```diff
@@ -2195,6 +2276,10 @@ protected RunInQueue inQueue() {
         return new RunInQueue(id, thdf, getHostSyncLevel());
     }
 
+    protected RunInQueue inQueue(int level) {
+        return new RunInQueue(id, thdf, level);
+    }
```
**Comment from tao.gan:**

Commit/Pull/Take/Merge 均采用 KVMGlobalConfig.HOST_SNAPSHOT_SYNC_LEVEL.value(Integer.class) 同步级别

MatheMatrix commented on plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageKvmBackend.java:
```diff
@@ -3864,4 +3868,47 @@ public void fail(ErrorCode errorCode) {
         return String.format("%s/%s-initialized-file", self.getMountPath(), self.getUuid());
     }
+
+    @Override
+    void handle(CommitVolumeSnapshotOnPrimaryStorageMsg msg, String hostUuid, final ReturnValueCompletion<CommitVolumeSnapshotOnPrimaryStorageReply> completion) {
+        CommitVolumeSnapshotOnPrimaryStorageReply reply = new CommitVolumeSnapshotOnPrimaryStorageReply();
+        OfflineCommitSnapshotCmd cmd = new OfflineCommitSnapshotCmd();
+        cmd.top = msg.getSrcSnapshot().getPrimaryStorageInstallPath();
+        cmd.base = msg.getDstSnapshot().getPrimaryStorageInstallPath();
+        cmd.topChildrenInstallPathInDb = msg.getSrcChildrenInstallPathInDb();
+        httpCall(OFFLINE_COMMIT_PATH, hostUuid, cmd, OfflineCommitSnapshotRsp.class, new ReturnValueCompletion<OfflineCommitSnapshotRsp>(completion) {
+            @Override
+            public void success(OfflineCommitSnapshotRsp returnValue) {
+                reply.setSize(returnValue.getActualSize());
+                completion.success(reply);
+            }
+
+            @Override
+            public void fail(ErrorCode errorCode) {
+                completion.fail(errorCode);
+            }
+        });
+    }
```
**Comment from tao.gan:**

已修改，LocalStorageKvmBackend.OfflineCommitSnapshotRsp 返回 long actualSize

zstack-robot-1 commented on plugin/nfsPrimaryStorage/src/main/java/org/zstack/storage/primary/nfs/NfsPrimaryStorage.java:
```diff
@@ -1936,4 +1881,92 @@ public void fail(ErrorCode errorCode) {
+        }
+
+        return hostUuid;
+    }
```
**Comment from tao.gan:**

已按照建议修改
增加 VolumeVO vol 和 List<HostInventory> connectedHosts 判空

zstack-robot-2 commented on testlib/src/main/java/org/zstack/testlib/vfs/storage/LocalStorageVFSPrimaryStorageTakeSnapshotBackend.groovy:
```diff
@@ -117,6 +116,34 @@ class LocalStorageVFSPrimaryStorageTakeSnapshotBackend implements AbstractFileSy
     }
+
+    @Override
+    Qcow2 blockCommit(HttpEntity<String> e, EnvSpec spec, KVMAgentCommands.BlockCommitCmd cmd, VolumeInventory volume) {
+        String storagePath = Q.New(PrimaryStorageVO.class)
+                .select(PrimaryStorageVO_.mountPath)
+                .eq(PrimaryStorageVO_.uuid, volume.primaryStorageUuid)
+                .findValue()
+        VFS vfs = LocalStorageSpec.vfs(LocalStorageSpec.hostUuidFromHTTPHeaders(e), storagePath, spec)
+        Qcow2 top = vfs.getFile(cmd.top, true)
+        Qcow2 base = vfs.getFile(cmd.base, true)
+        Qcow2.commit(vfs, top, base)
+        return vfs.getFile(cmd.base, true)
+    }
```
**Comment from tao.gan:**

`vfs.getFile(path) 如果path不存储，会直接报错，不会默认创建占位`

MatheMatrix commented on testlib/src/main/java/org/zstack/testlib/LocalStorageSpec.groovy:
```diff
@@ -508,11 +509,43 @@ class LocalStorageSpec extends PrimaryStorageSpec {
+                }
+                rsp.actualSize = 1
+                return rsp
+            }
```
**Comment from tao.gan:**

已增加

zstack-robot-2 commented on storage/src/main/java/org/zstack/storage/snapshot/VolumeTree.java:
```diff
@@ -0,0 +1,546 @@
+                    // update distance
+                    vo.setDistance(vo.getDistance() - 1);
+
+                    if (Objects.equals(vo.getUuid(), dstSnapshotInv.getUuid())) {
+                        // update parentUuid
+                        vo.setParentUuid(srcSnapshotInv.getParentUuid());
+                        logger.debug(String.format("update the parent of snapshot[uuid:%s] to %s", dstSnapshotInv.getUuid(), srcSnapshotInv.getParentUuid()));
+
+                        vo.setSize(newInstallPathSize);
+                        logger.debug(String.format("update the size of snapshot[uuid:%s] to %s", dstSnapshotInv.getUuid(), newInstallPathSize));
+                    }
+                });
+
+                VolumeSnapshotTreeVO newTree = null;
+                // create new tree and update treeUuid
+                if (srcSnapshotInv.getParentUuid() == null) {
+                    newTree = new VolumeSnapshotTreeVO();
+                    newTree.setCurrent(descendantsUuid.contains(volume.getUuid()));
+                    newTree.setVolumeUuid(volume.getUuid());
+                    newTree.setUuid(Platform.getUuid());
+                    newTree.setStatus(VolumeSnapshotTreeStatus.Completed);
+                    if (getAliveChainSnapshotUuids().contains(dstSnapshotInv.getUuid())) {
+                        newTree.setCurrent(true);
+                    }
+                    dbf.persist(newTree);
+                    logger.debug(String.format("created new volume snapshot tree[uuid:%s]", newTree.getUuid()));
+                    VolumeSnapshotTreeVO finalNewTree = newTree;
+                    vos.forEach(vo -> vo.setTreeUuid(finalNewTree.getUuid()));
+                }
+
+                dbf.updateCollection(vos);
+
+                if (newTree != null && Objects.equals(dstSnapshotInv.getUuid(), volume.getUuid())
+                        && q(VolumeSnapshotTreeVO.class).eq(VolumeSnapshotTreeVO_.uuid, srcSnapshotInv.getTreeUuid()).count() == 1) {
+                    sql(VolumeSnapshotTreeVO.class).eq(VolumeSnapshotTreeVO_.uuid, srcSnapshotInv.getTreeUuid())
+                            .set(VolumeSnapshotTreeVO_.current, false).update();
+                }
+            }
+        }.execute();
+    }
```
**Comment from tao.gan:**

核心逻辑，暂时不重构

MatheMatrix commented on storage/src/main/java/org/zstack/storage/snapshot/VolumeTree.java:
```diff
@@ -0,0 +1,546 @@
+        return aliveChain.stream().map(VolumeSnapshotInventory::getUuid).collect(Collectors.toList());
+    }
+
+    public DeleteVolumeSnapshotDirection resolveDirection(String targetSnapshotUuid, String childSnapshotUuid, String initialDirection,
+                                                          boolean targetSnapshotIsLatest, VmInstanceState vmState) {
+        boolean online = (vmState == VmInstanceState.Running || vmState == VmInstanceState.Paused)
+                && getAliveChainSnapshotUuids().contains(targetSnapshotUuid) && getAliveChainSnapshotUuids().contains(childSnapshotUuid);
+
+        boolean shouldUseCommitStrategy = current && !targetSnapshotIsLatest && online;
+
+        if (Objects.equals(initialDirection, DeleteVolumeSnapshotDirection.Pull.toString()) && shouldUseCommitStrategy) {
+            throw new IllegalArgumentException("the snapshot will be deleted by block 'commit', but the direction is 'pull', " +
+                    "change the direction to 'commit' or 'auto'.");
+        }
+
+        if (initialDirection == null) {
+            return DeleteVolumeSnapshotDirection.Commit;
+        }
+
+        if (Objects.equals(initialDirection, DeleteVolumeSnapshotDirection.Auto.toString())) {
+            return shouldUseCommitStrategy ?
+                    DeleteVolumeSnapshotDirection.Commit :
+                    DeleteVolumeSnapshotDirection.Pull;
+        }
+
+        return DeleteVolumeSnapshotDirection.fromString(initialDirection);
+    }
```
**Comment from tao.gan:**

`initialDirection` 使用方式正确，不用修改

zstack-robot-1 commented on storage/src/main/java/org/zstack/storage/snapshot/VolumeSnapshotTreeBase.java:
```diff
@@ -738,16 +756,633 @@ public void fail(ErrorCode errorCode) {
+
+                            @Override
+                            public void run(FlowTrigger trigger, Map data) {
+                                SyncVolumeSizeMsg msg = new SyncVolumeSizeMsg();
+                                msg.setVolumeUuid(volume.getUuid());
+                                bus.makeTargetServiceIdByResourceUuid(msg, VolumeConstant.SERVICE_ID, currentRoot.getPrimaryStorageUuid());
+                                bus.send(msg, new CloudBusCallBack(trigger) {
+                                    @Override
+                                    public void run(MessageReply reply) {
```
**Comment from tao.gan:**

已修改
